### PR TITLE
Add end to end integration tests for external IdP

### PR DESCRIPTION
**Pull request by amore17**
Added tests for HBAC and SUDO rule and other test scenarios.
    
Related : https://pagure.io/freeipa/issue/8805
Related: https://pagure.io/freeipa/issue/8803
Related: https://pagure.io/freeipa/issue/8804

### CODE REVIEW

File: ipatests/test_integration/test_idp.py [hunk 1]
```diff
@@ -58,11 +61,11 @@ def kinit_idp(host, user, keycloak_server):
     tasks.kdestroy_all(host)
     # create armor for FAST
     host.run_command(["kinit", "-n", "-c", ARMOR])
-    since = time.strftime('%Y-%m-%d %H:%M:%S')
     cmd = ["kinit", "-T", ARMOR, user]
     with host.spawn_expect(cmd, default_timeout=100) as e:
         e.expect('Authenticate at .+: ')
-        uri = get_verification_uri(host, since, keycloak_server.hostname)
+        time.sleep(15)
```
Comment (flo-renaud) [inline ipatests/test_integration/test_idp.py:67]: Why this change? Was the test unstable?

Comment (amore17) [reply ipatests/test_integration/test_idp.py:67]: Its stable in PR-CI environment but its unstable in idm-ci environment machines are a bit faster.

File: ipatests/test_integration/test_idp.py [hunk 2]
```diff
@@ -74,43 +77,219 @@ def kinit_idp(host, user, keycloak_server):
 
 class TestIDPKeycloak(IntegrationTest):
 
-    num_replicas = 1
+    num_replicas = 2
```
[the remark this answers is not in the document]
Comment (amore17) [reply ipatests/test_integration/test_idp.py:80]: Yes it needs a big enough machine to host Keycloak server.

File: ipatests/test_integration/test_idp.py [hunk 3]
```diff
@@ -74,43 +77,219 @@ def kinit_idp(host, user, keycloak_server):
...
         self.master.run_command(cmd, stdin_text="{0}\n{0}".format(
             keycloak_srv.config.admin_password))
         tasks.user_add(self.master, 'keycloakuser',
                        extra_args=["--user-auth-type=idp",
                                    "--idp-user-id=testuser1@ipa.test",
-                                   "--idp=keycloak"]
+                                   "--idp=keycloakidp"]
                        )
+        tasks.user_add(self.master, 'testuser')
+        tasks.clear_sssd_cache(self.master)
```
[the remark this answers is not in the document]
Comment (amore17) [reply ipatests/test_integration/test_idp.py:132]: Removed.

File: ipatests/test_integration/test_idp.py [hunk 4]
```diff
@@ -74,43 +77,219 @@ def kinit_idp(host, user, keycloak_server):
...
+            ['ipa', 'sudorule-add-user', '--users=keycloakuser',
+             '--groups=admins', 'sudorule'],
+            ['ipa', 'sudorule-add-host', '--hosts',
+             client.hostname, 'sudorule'],
+            ['ipa', 'sudorule-add-runasuser',
+             '--users=root', 'sudorule'],
+            ['ipa', 'sudorule-add-allow-command',
+             '--sudocmds=/usr/bin/yum', 'sudorule'],
+            ['ipa', 'sudorule-show', 'sudorule', '--all']
+        ]
```
[the remark this answers is not in the document]
Comment (amore17) [reply ipatests/test_integration/test_idp.py:199]: Removed admins.

File: ipatests/test_integration/test_idp.py [hunk 5]
```diff
@@ -74,43 +77,219 @@ def kinit_idp(host, user, keycloak_server):
...
+        services = ["DNS/{0}@{1}".format(master.hostname, domain),
+                    "HTTP/{0}@{1}".format(client.hostname, domain),
+                    "dogtag/{0}@{1}".format(master.hostname, domain),
+                    "ipa-dnskeysyncd/{0}@{1}".format(master.hostname, domain)
+                    ]
+        for service in services:
+            test = master.run_command(["ipa", "service-mod", service,
+                                       "--auth-ind=idp"]
+                                      )
+            assert "Authentication Indicators: idp" in test.stdout_text
```
[the remark this answers is not in the document]
Comment (amore17) [reply ipatests/test_integration/test_idp.py]: Fixed.

File: ipatests/test_integration/test_idp.py [hunk 6]
```diff
@@ -58,11 +62,11 @@ def kinit_idp(host, user, keycloak_server):
     tasks.kdestroy_all(host)
     # create armor for FAST
     host.run_command(["kinit", "-n", "-c", ARMOR])
-    since = time.strftime('%Y-%m-%d %H:%M:%S')
     cmd = ["kinit", "-T", ARMOR, user]
     with host.spawn_expect(cmd, default_timeout=100) as e:
         e.expect('Authenticate at .+: ')
-        uri = get_verification_uri(host, since, keycloak_server.hostname)
+        time.sleep(15)
+        uri = get_verification_uri(host, keycloak_server.hostname)
```
Comment (flo-renaud) [inline ipatests/test_integration/test_idp.py:69]: The `get_verification_uri` seems a complex way of retrieving the URI, by reading the content of the journal.
The code just above, `e.expect('Authenticate at .+: ')` could probably be  used to retrieve the URI, as the output is similar to `Authenticate at https://replica1.ipa.test:8443/auth/realms/master/device?user_code=OWAM-ZOPR and press ENTER.: `. I believe it's possible to extract the url directly from the output using e.match.groups() if you change e.expect to something like e.expect(Authenticate at (.+) and press ENTER.:`)

**Review by flo-renaud — COMMENTED** [review]
Hi @amore17 
thanks for the update. Please find my inline comments + the following: the selenium driver produces a deprecation warning:
```
/tmp/add_user_code.py:10: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  driver = webdriver.Firefox(executable_path="/opt/geckodriver", options=options)
/tmp/add_user_code.py:16: DeprecationWarning: find_element_by_* commands are deprecated. Please use find_element() instead
```
Not urgent to fix but we may hit an issue with the next version of the driver.

File: ipatests/test_integration/test_idp.py [hunk 7]
```diff
@@ -74,43 +78,222 @@ def kinit_idp(host, user, keycloak_server):
...
 
     @classmethod
     def install(cls, mh):
-        tasks.install_master(cls.master, setup_dns=True)
-        tasks.install_client(cls.master, cls.replicas[0])
-        content = cls.master.get_file_contents(paths.IPA_DEFAULT_CONF,
-                                               encoding='utf-8')
-        new_content = content + "\noidc_child_debug_level = 10"
-        cls.master.put_file_contents(paths.IPA_DEFAULT_CONF, new_content)
+        cls.master = cls.master
```
[the remark this answers is not in the document]
Comment (flo-renaud) [reply ipatests/test_integration/test_idp.py:81]: I can't understand this line: `cls.master = cls.master`

**Review by flo-renaud — COMMENTED** [review]
Hi @amore17 
thanks for the PR.
Could you split the commit into 2 different commits, one containing the prci definitions changes and the other the diff for test_idp.py? For downstream backports, we don't pick the patches modifying the yaml files and having them in a separate patch makes it easier.
Thanks

File: ipatests/test_integration/test_idp.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (amore17) [reply ipatests/test_integration/test_idp.py]: It was added by mistaken. Removed.

**Comment by amore17** [comment]
temp_commit [logs](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/3fba9592-ee5b-11ec-adda-fa163eab4448)

**Comment by amore17** [comment]
> Hi @amore17 thanks for the PR. Could you split the commit into 2 different commits, one containing the prci definitions changes and the other the diff for test_idp.py? For downstream backports, we don't pick the patches modifying the yaml files and having them in a separate patch makes it easier. Thanks

Added changes.

File: ipatests/test_integration/test_idp.py [hunk 8]
```diff
@@ -97,20 +102,199 @@ def install(cls, mh):
...
+            self.master.run_command(cmd)
+        tasks.clear_sssd_cache(self.master)
+        tasks.clear_sssd_cache(self.client)
+        try:
+            cmd = 'sudo -ll -U keycloakuser'
+            test = self.client.run_command(cmd).stdout_text
+            assert "User keycloakuser may run the following commands" in test
+            assert "/usr/bin/yum" in test
+            list_fail = self.master.run_command(cmd).stdout_text
+            assert "User keycloakuser is not allowed to run sudo" in list_fail
```
Comment (flo-renaud) [inline ipatests/test_integration/test_idp.py:209]: I would also try to run the `sudo yum` command as the user `keycloakuser` after authenticating with the external IdP.

File: ipatests/test_integration/test_idp.py [hunk 9]
```diff
@@ -97,20 +102,199 @@ def install(cls, mh):
...
+        try:
+            list_user = self.master.run_command(
+                ['ipa', 'user-show', 'backupuser', '--all']
+            ).stdout_text
+            assert "External IdP configuration: testidp" in list_user
+            assert "User authentication types: idp" in list_user
+            assert ("External IdP user identifier: "
+                    "testuser1@ipa.test") in list_user
+            list_idp = self.master.run_command(['ipa', 'idp-find', 'testidp'])
+            assert "testidp" in list_idp.stdout_text
```
Comment (flo-renaud) [inline ipatests/test_integration/test_idp.py:297]: I would also check that authentication with external IdP is also working after the restore.

**Review by flo-renaud — COMMENTED** [review]
Hi @amore17 
please find inline comments.

File: ipatests/test_integration/test_idp.py [hunk 10]
```diff
@@ -97,20 +102,208 @@ def install(cls, mh):
...
+            self.master.run_command(cmd)
+        tasks.clear_sssd_cache(self.master)
+        tasks.clear_sssd_cache(self.client)
+        try:
+            cmd = 'sudo -ll -U keycloakuser'
+            test = self.client.run_command(cmd).stdout_text
+            assert "User keycloakuser may run the following commands" in test
+            assert "/usr/bin/yum" in test
+            kinit_idp(self.client, 'keycloakuser', self.client)
+            test_sudo = 'sudo -u root yum remove wget'
```
Comment (flo-renaud) [inline ipatests/test_integration/test_idp.py:211]: 1. the yum remove command is cancelled by the No prompt, and returns a non-0 exit status. You can use a simple `yum list wget` command instead.
2. with this command, it's the root user who is running sudo -u root. The command aims to test if keycloak user can run sudo, it should look like:
3. `test_sudo = 'su -c "sudo yum list wget" keycloakuser`

**Review by flo-renaud — COMMENTED** [review]
Hi @amore17 
please find one inline comment

File: ipatests/test_integration/test_idp.py (hunk 10 above)
Comment (amore17) [reply ipatests/test_integration/test_idp.py]: Added changes.

**Comment by flo-renaud** [comment]
For future ref, link to successful run of test_idp: [Details](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/82c056c0-f201-11ec-99c8-fa163e04db38)

**Comment by flo-renaud** [comment]
master:

* a80a98194837f99c494dbee7e3326acde1ccb8cd ipatests: update prci definitions for test_idp.py
* bd57ff35660cd1e911db43805fbbea5d26058fbc Add end to end integration tests for external IdP

**Comment by flo-renaud** [comment]
Hi @amore17 
can you do a manual backport to ipa-4-9? the test definitions need to be adapted on this branch. Thanks
